pczt/src/roles/signer/mod.rs

```diff
@@ -0,0 +1,302 @@
+    ///
+    /// It is the caller's responsibility to perform any semantic validity checks on the
+    /// PCZT (for example, comfirming that the change amounts are correct) before calling
+    /// this method.
+    pub fn sign_transparent(
+        &mut self,
+        index: usize,
+        sk: &secp256k1::SecretKey,
+    ) -> Result<(), Error> {
+        let input = self
+            .transparent
+            .inputs_mut()
+            .get_mut(index)
+            .ok_or(Error::InvalidIndex)?;
+
+        // Check consistency of the input being signed.
+        // TODO
+
+        input
+            .sign(index, &self.tx_data, &self.txid_parts, sk, &self.secp)
+            .map_err(Error::TransparentSign)?;
+
+        // Update transaction modifiability.
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !0b0000_0001;
+        }
+        if ![SighashType::NONE, SighashType::NONE_ANYONECANPAY].contains(input.sighash_type()) {
+            self.global.tx_modifiable &= !0b0000_0010;
+        }
+        if [SighashType::SINGLE, SighashType::SINGLE].contains(input.sighash_type()) {
+            self.global.tx_modifiable |= 0b0000_0100;
+        }
+
+        Ok(())
+    }
+
+    /// Signs the Sapling spend at the given index with the given spend authorizing key.
+    ///
+    /// It is the caller's responsibility to perform any semantic validity checks on the
+    /// PCZT (for example, comfirming that the change amounts are correct) before calling
+    /// this method.
+    pub fn sign_sapling(
+        &mut self,
+        index: usize,
+        ask: &sapling::keys::SpendAuthorizingKey,
+    ) -> Result<(), Error> {
+        let spend = self
+            .sapling
+            .spends_mut()
+            .get_mut(index)
+            .ok_or(Error::InvalidIndex)?;
+
+        // Check consistency of the input being signed.
+        let note_from_fields = spend
+            .recipient()
+            .zip(spend.value().as_ref())
+            .zip(spend.rseed().as_ref())
+            .map(|((recipient, value), rseed)| {
+                sapling::Note::from_parts(recipient, *value, *rseed)
+            });
+
+        if let Some(note) = note_from_fields {
+            let tx_spend = self
+                .tx_data
+                .sapling_bundle()
+                .expect("index checked above")
+                .shielded_spends()
+                .get(index)
+                .expect("index checked above");
+
+            let proof_generation_key = spend
+                .proof_generation_key()
+                .as_ref()
+                .ok_or(Error::MissingProofGenerationKey)?;
+
+            let nk = proof_generation_key.to_viewing_key().nk;
+
+            let merkle_path = spend.witness().as_ref().ok_or(Error::MissingWitness)?;
+
+            if &note.nf(&nk, merkle_path.position().into()) != tx_spend.nullifier() {
```

Comment:
I thought that this might be imposing an unnecessary constraint by requiring the witness of the spent note rather than just the position. But actually the constraint is necessary, because the sighash for Sapling spend auth signatures is over the anchor [even in v5 transactions](https://zips.z.cash/zip-0244#t-3a-ii-sapling-spends-noncompact-digest). And so even if the anchor is made optional [as I suggested](https://github.com/zcash/librustzcash/pull/1577/files#r1873734850), we cannot update it after making signatures. (The sighash for Orchard spend auth signatures is [also over the anchor](https://zips.z.cash/zip-0244#t-4-orchard-digest).)